contrib/bindings/go/pathrs/pathrs.go

```diff
@@ -0,0 +1,328 @@
+	handler := C.pathrs_creat(r.root, C.CString(path), C.uint(mode))
+	if handler == nil {
+		return nil, handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+	}
+
+	return newHandle(handler), nil
+}
+
+// Rename Within the given root's tree, perform the rename of src to dst,
+// or change flags on this file if the names are the same it's only change the flags
+func (r *Root) Rename(src, dst string, flags int) error {
+	C.pathrs_rename(r.root, C.CString(src), C.CString(dst), C.int(flags))
+	return handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+}
+
+// Mkdir creates a directory with a such mode by path
+func (r *Root) Mkdir(path string, mode uint) error {
+	C.pathrs_mkdir(r.root, C.CString(path), C.uint(mode))
+	return handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+}
+
+// Mknod creates a filesystem node named path
+// with attributes mode and dev
+func (r *Root) Mknod(path string, mode uint, dev int) error {
+	C.pathrs_mknod(r.root, C.CString(path), C.uint(mode), C.dev_t(dev))
+	return handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+}
+
+// Hardlink creates a hardlink of file named target and place it to path
+func (r *Root) Hardlink(path, target string) error {
+	C.pathrs_hardlink(r.root, C.CString(path), C.CString(target))
+	return handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+}
+
+// Symlink creates a symlink of file named target and place it to path
+func (r *Root) Symlink(path, target string) error {
+	C.pathrs_symlink(r.root, C.CString(path), C.CString(target))
+	return handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+}
+
+// IntoFd unwraps a file-based libpathrs object to obtain its underlying file
+// descriptor.
+//
+// It is critical that you do not operate on this file descriptor yourself,
+// because the security properties of libpathrs depend on users doing all
+// relevant filesystem operations through libpathrs.
+func (r *Root) IntoFd() (int, error) {
+	fd := int(C.pathrs_into_fd(C.PATHRS_ROOT, unsafe.Pointer(r.root)))
+	if fd < 0 {
+		return 0, handleErr(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+	}
+
+	return fd, nil
+}
+
+// Clone creates a copy of root handler the new object will have a separate lifetime
+// from the original, but will refer to the same underlying file
+func (r *Root) Clone() (*Root, error) {
+	newRoot := (*C.pathrs_root_t)(C.pathrs_duplicate(C.PATHRS_ROOT, unsafe.Pointer(r.root)))
+	err := handleErr(C.PATHRS_ROOT, unsafe.Pointer(newRoot))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Root{root: newRoot}, nil
+}
+
+// Free frees underling caught resources
+func (r *Root) Free() {
+	if r != nil {
+		C.pathrs_free(C.PATHRS_ROOT, unsafe.Pointer(r.root))
+	}
+}
+
+func newHandle(handlePtr *C.pathrs_handle_t) *Handle {
+	return &Handle{
+		handle:  handlePtr,
+		nameGen: rand.New(rand.NewSource(time.Now().UTC().UnixNano())),
+	}
+}
+
+// Reopen upgrade the handle to a file representation
+// which holds a usable fd, suitable for reading and writing
+func (h *Handle) Reopen(flags int) (*os.File, error) {
```

Comment:
I don't like that callers would have to pass `unix.O_RDONLY` (it's unlike most Go libraries)-- ideally there would be separate methods for each (to match `os.Open` and `os.CreateFile`). But I can fix this later...

Reply:
Might it makes sense to provide a highly similar api to the `os` package?

This is a snippet that demonstrates the idea. In case if someone has some specifics he/she can use the standard api with `Root` and `Handle`

```

type File struct {
	root *Root
	handle *Handle
	*os.File
}

func Open(root, path string) (*File, error) {
	return OpenFile(root, path, os.O_RDONLY)
}

func OpenFile(rootPath, path string, flags int) (*File, error) {
	root, err := RootOpen(rootPath)
	if err != nil {
		return nil, err
	}

	handle, err := root.Resolve(path)
	if err != nil {
		return nil, err
	}

	file, err := handle.OpenWith(flags)
	if err != nil {
		return nil, err
	}

	return &File {
		root,
		handle,
		file,
	}, nil
}

func (f *File) Close() {
	f.root.Close()
	f.handle.Close()
	f.File.Close()
}

```

Reply:
There should probably be a short-hand for `pathrs_resolve`+`pathrs_reopen`, but I don't like the idea of having an interface which is basically `open(root, path)` for a few reasons:

 * It's not type-safe, so if you get the argument order wrong you've just opened an attacker-controlled path. Yeah, `Open` has the same problem but ideally you'll be creating one `Root` object for all operations within a given root.

 * It looks too much like `filepath.Join` which makes me worried since the entire point of `libpathrs` is that using string paths is the wrong interface for paths in certain applications. Encouraging people to use string paths is a recipe for disaster IMHO (it'll be a few days before someone starts doing `filepath.Join` again out of habit).

 * I don't like that `File` contains all three objects (the root, handle, and re-opened file). A "file handle" shouldn't care about what root it came from.

If someone really wants to do something like that, they can just write an implementation in their own packages.

Reply:
You are right, thank you.